src/firmware/app/primitives/direct_control_primitive.c

```diff
@@ -0,0 +1,141 @@
+                                 control_msg.back_right_wheel_rpm);
+            app_wheel_applyForce(app_firmware_robot_getFrontRightWheel(robot),
+                                 control_msg.front_right_wheel_rpm);
+            break;
+        }
+        case DirectControlPrimitiveMsg_direct_velocity_control_tag:
+        {
+            DirectControlPrimitiveMsg_DirectVelocityControlMsg control_msg =
+                prim_msg.wheel_control.direct_velocity_control;
+            state->direct_velocity          = true;
+            state->direct_target_velocity_x = control_msg.velocity.x_component_meters;
+            state->direct_target_velocity_y = control_msg.velocity.y_component_meters;
+            state->direct_target_velocity_angular =
+                control_msg.angular_velocity.radians_per_second;
+            break;
+        }
+        default:
+        {
+            // Do nothing
+            // TODO (Issue #1652): LOG(WARNING)<<"Wheel control command is not a valid
+            // type"
+            state->direct_velocity = false;
+        }
+    }
+
+    switch (prim_msg.charge_mode)
+    {
+        case DirectControlPrimitiveMsg_ChargeMode_CHARGE:
+        {
+            app_charger_charge_capacitor(charger);
+            break;
+        }
+        case DirectControlPrimitiveMsg_ChargeMode_FLOAT:
+        {
+            app_charger_float_capacitor(charger);
+            break;
+        }
+        default:
+        {
+            // discharge
+            app_chicker_disableAutochip(chicker);
+            app_chicker_disableAutokick(chicker);
+            app_charger_discharge_capacitor(charger);
+        }
+    }
+
+    switch (prim_msg.which_chick_command)
+    {
+        case DirectControlPrimitiveMsg_kick_speed_meters_per_second_tag:
+        {
+            app_chicker_kick(chicker,
+                             prim_msg.chick_command.kick_speed_meters_per_second);
+            break;
+        }
+        case DirectControlPrimitiveMsg_chip_distance_meters_tag:
+        {
+            app_chicker_chip(chicker, prim_msg.chick_command.chip_distance_meters);
+            break;
+        }
+        case DirectControlPrimitiveMsg_autokick_speed_meters_per_second_tag:
+        {
+            app_chicker_enableAutokick(
+                chicker, prim_msg.chick_command.autokick_speed_meters_per_second);
+            break;
+        }
+        case DirectControlPrimitiveMsg_autochip_distance_meters_tag:
+        {
+            app_chicker_enableAutochip(chicker,
+                                       prim_msg.chick_command.autochip_distance_meters);
+            break;
+        }
+        default:
+        {
+            // Do nothing
+            // TODO (Issue #1652): LOG(WARNING)<<"Chick command is not a valid type"
+        }
+    }
+
+    app_dribbler_setSpeed(dribbler, (uint32_t)prim_msg.dribbler_speed_rpm);
+}
+
+static void direct_control_end(void* void_state_ptr, FirmwareWorld_t* world) {}
```

Comment:
should we be zeroing out all the velocities on the dribbler and and wheels and maybe discharging caps here?

Reply:
`app_primitive_manager_makeRobotSafe` should do the trick